### DISCUSSION TITLE
How do I update query/auth on connect?

### DISCUSSION THREAD

**Discussion by FerusAndBeyond** [socketio/socket.io#3902] [Q&A] [answered]
How can query/auth be updated during reconnect() or connect() to a new namespace? Whenever I try to do this the initially set auth/query is kept:

```
io(URL, { auth: { token } })
// or
io(URL, { query: { token } })
```

If I set token to `null` initially, then in the next namespace I connect to the token will also be null even if I change it. I'm using `"socket.io-client": "^4.0.1",`

**Comment by darrachequesne** [comment, marked as answer]
You can update those options in the "reconnect_attempt" event:

```js
socket.io.on("reconnect_attempt", () => {
  socket.io.opts.query.token = 123;
  socket.auth.token = 456;
});
```

Please note that the `query` option will be shared between the Socket instances of the same connection.

Reference:
- https://socket.io/docs/v4/client-initialization/#query
- https://socket.io/docs/v4/client-initialization/#auth

**Reply by FerusAndBeyond** [reply]
Thank you it seems to work. I have another question. What happens when a namespace-socket is disconnected?

First I connect to the socket in a namespace

```
const sio = io(url, {
                transports: ["websocket", "polling"],
                query: {
                    token: auth.token
                }
            })
```

Then later I can manually disconnect from the namespace using:
```
sio.io.disconnect()
sio.close()
```
 When I try to reconnect again using `sio.io.connect()` a connect-event is not sent to the server. I also tried

```
sio.io.on("reconnect", function () {
    sio.emit("reconnecting")
});
```

But the reconnecting event is only _sometimes_ sent, so not sure what is going on.

**Reply by FerusAndBeyond** [reply]
Okay I think I made it work by changing `sio.io.connect()` to `sio.connect()`. What is the difference between the two?

**Reply by darrachequesne** [reply]
The `io` object points to the Manager instance, which manages the low-level connection.

The `sio` object in your example is the Socket instance. A single Manager can be shared by multiple Socket instances (one per namespace).

```
sio.io.disconnect(); // closes the underlying connection (so all Socket instances disconnect)

sio.close(); // disconnect the given Socket instance (but the Manager may not be closed, if there are other active Socket instances) 
```

**Reply by FerusAndBeyond** [reply]
I see. It doesn't seem to work without `sio.io.disconnect()`,  otherwise no reconnection is made.
